### DISCUSSION TITLE
Options to seed a database with significant amount of data

### DISCUSSION THREAD

**Discussion by ChristianIvicevic** [planetscale/discussion#471] [Q&A] [answered]
I am working on a small personal project tracking card collections for a TCG. For that I am working on a set of (read-only) tables that contain information about all available cards while only writing to some other tables to track the inventory of users. The tables containing all cards will never be written to from the application and only manually updated once new sets and cards get released.

A full export into a raw SQL file is about 410MB with approx 1.7M lines of commands. For testing purposes I only picked a small subset and attempted to seed the database on Planetscale with a file of about 20MB with approx 80k lines of commands. This process took 35-40 minutes and indicates that an initial seed of the database will take 6-12h. You may imagine that I am afraid to trigger this process and lose connection halfway through the process etc.

What options are there to reliably seed a database with this much data?

**Comment by orware** [comment]
Hello @ChristianIvicevic,

Could you provide an example of how your SQL file currently looks and what procedure is being used for loading it into your database to provide some potential clues we can share for you to possibly accomplish this more quickly/reliably?

We can also potentially work more closely together via ticket if that's preferred too over at: https://support.planetscale.com/.

**Reply by ChristianIvicevic** [reply]
I can share the SQL file itself, sure. You'll find the dumped file over here under _All Printings_: https://mtgjson.com/downloads/all-files/
![image](https://github.com/planetscale/discussion/assets/45569997/1973942d-ca1b-4d63-bf1d-2aa8fa689ef7)

Prior to importing it via `pscale shell < ./AllPrintings.sql` I manually made changes to remove the foreign key constraints in the `CREATE TABLE` commands to ensure compatibility with Planetscale. Let me know if you need more information.

**Reply by orware** [reply]
Thanks @ChristianIvicevic!

I was able to grab the file and have it going with an import after completing that same manual step you described about removing the foreign key constraints sections of the table definitions.

I'm going to leave my import running to see how long it may take to complete but it is running relatively slow right now as you described above.

In my review of the file itself, it appears to have the `INSERT` queries organized in such a way that everything is individualized, with a query for each `INSERT`, along with having rows for different tables interspersed throughout the file.

I think that combination likely leads to the poor performance at the moment, although I don't know if you have any options for regenerating or re-exporting the files from the original source here using different options.

I do see that you have CSV files as well, which you may be able to use and import using a program like DBeaver, via it's [CSV import functionality](https://github.com/planetscale/discussion/discussions/372), but I'd likely want to run through that set of tests myself too in order to verify if it is truly faster so I'll probably start that separately here in a bit. Of course, DataGrip probably has a feature for that too ;-).

Aside from that, a `.sql` file that contains the inserts already arranged in batches of 50 to 100 would probably perform a bit better where the inserts are in a format that more or less looks like this:

```sql
INSERT INTO example(id, name) VALUES (1, 'Christian'), (2, 'Omar'), ...,  (50, 'Bob');
```

Since that would lower the overall amount of queries the database would need to process as it goes through the file.

I'll come back and share some general timing results once my tests complete, but currently I would think the batch inserts approach, or possibly the CSV import approach, will probably end up being the fastest ones. The only issue with the batch insert approach would be needing to recreate the `.sql` files in that way.

**Reply by orware** [reply]
One other observation I forgot to mention:

As I was reviewing the schema included in the file, I saw that there were a lot of `TEXT` fields present, but I'm unsure if the tables actually require storing `64 KB` of text in each of those fields. If not, switching to a longer `VARCHAR` field type is generally recommended for improved performance too.

**Reply by ChristianIvicevic** [reply]
Technically all the data originates from one or more JSON files without strict schema. There is an open source script the team behind that particular project uses to manually build the SQL dump and a fully working SQLite file. As such I am technically able to locally derive a set of commands with a custom script that may be more optimized depending on the solution you propose.

**Reply by ChristianIvicevic** [reply]
> If not, switching to a longer `VARCHAR` field type is generally recommended for improved performance too.

Indeed, the derived SQL schema is fairly inefficient and bad. Most obvious caveat is that all those `TEXT` can't be trivially indexed, so this is definitely another step when I attempt to process the original schema.

**Reply by orware** [reply]
I'm going to try a round of import testing using the CSV files, but here is the direct result using the `AllPrintings.sql` file and importing that file directly via `pscale shell`:

![image](https://github.com/planetscale/discussion/assets/931650/bb826a94-aa5d-4083-9636-8dede1a85b9b)

The overall time was about 19.5 hours. 

A lot of that time spent is likely due to the inefficiency of the way the inserts are being performed currently in the `AllPrintings.sql` file that I mentioned above from yesterday, plus also a bit due to being limited by the Hobby plan database resources vs. a paid database, but I believe you are also on a Hobby plan database so at least it's more of an apples-to-apples comparison too.

I'll update again once I have the CSV import method test results to share for comparison!

**Reply by orware** [reply, marked as answer]
@ChristianIvicevic,

The CSV import results using DBeaver were considerably quicker in comparison to the first import test above and only really took minutes in comparison and the most likely factor here was just the batch size used.

![image](https://github.com/planetscale/discussion/assets/931650/8787fe41-9931-4695-86b1-815ad1d8c811)

Total import time for this was about ~30 minutes and the main delays were figuring out the DBeaver settings to use and making sure the "index" column included in the CSV files was excluded/skipped since it wasn't actually part of the table columns in the schema. The main drawback with this approach was simply needing to import each CSV file individually for each table.

Below are the example settings I used within DBeaver's "Data load settings" screen as part of the "Import Data" wizard mentioned in [discussions#372](https://github.com/planetscale/discussion/discussions/372):
![image](https://github.com/planetscale/discussion/assets/931650/85b4563a-855c-42a5-8002-aeafb2fc6e33)

Given the above results, reworking the SQL files so that the inserts are already arranged in batches of 100 should lead to a considerably faster import for you compared to the current arrangement in the `AllPrintings.sql` file that is only using individual insert queries.

**Reply by ChristianIvicevic** [reply]
Absolutely fantastic - appreciate the effort here! I would have most likely reached some limits just attempting to experiment with different options.

Moving forward I will most likely grab the published SQLite or CSVs, map it onto a custom CSV with a better schema and seed it using Data Grip. In fact DG allows to select multiple CSV files at the same time, however I haven't found an option to deal with batches at a first glance. Will test with a smaller data set to estimate whether this impacts performance significantly or not. And if it does, I'll just let the DG team know - the advantages of working at JetBrains myself 😄 

Once again, thanks for the investigation on this topic!

**Reply by ChristianIvicevic** [reply]
@orware I did a few different imports with DataGrip. Importing the entire `cards.csv` without automatic detection of the schema and no optimization whatsoever took about 2min and 15s for all 80.000 rows. DG seems to batch 1.000 rows at once per insert according the Insights tab. This is an absolutely great performance for such a badly prepared data set and this will end up taking only a fraction of the time once I design a better schema and only pick a subset of the original columns.

Definitely worth working with CSV for seeding database on Planetscale.
